### PR TITLE
91 senet

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Implemented #91  .

### Description
SENet, SEResnet and SEResneXt have been implemented. In order to make full use of MONAI's modules, some places have been modified in convolutions.py and squeeze_and_excitation.py, and a few typing hints have been added into unet.py.
As for the integration tests, there are some errors (see below) before I modify the code, thus I am not sure if I should do something for that right now.

<pre>monai/networks/utils.py:73: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/networks/utils.py:135: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/networks/utils.py:137: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/networks/layers/simplelayers.py:100: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/focal_loss.py:87: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/focal_loss.py:89: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/focal_loss.py:105: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/focal_loss.py:119: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/focal_loss.py:127: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/rocauc.py:101: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/rocauc.py:104: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/rocauc.py:114: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/meandice.py:162: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/meandice.py:191: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/metrics/meandice.py:192: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/tversky.py:90: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/dice.py:95: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/dice.py:171: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/losses/dice.py:251: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/inferers/utils.py:72: <font color="#CC0000"><b>error:</b></font> Need type annotation for &apos;image_size_&apos; (hint: <b>&quot;image_size_: List[&lt;type&gt;] = ...&quot;</b>)
monai/transforms/post/array.py:59: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/transforms/post/array.py:335: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/transforms/post/array.py:419: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
monai/networks/blocks/squeeze_and_excitation.py:75: <font color="#CC0000"><b>error:</b></font> Need more than 1 value to unpack (2 expected)
monai/networks/blocks/squeeze_and_excitation.py:76: <font color="#CC0000"><b>error:</b></font> Cannot determine type of &apos;b&apos;
monai/networks/blocks/squeeze_and_excitation.py:76: <font color="#CC0000"><b>error:</b></font> Cannot determine type of &apos;c&apos;
monai/networks/blocks/squeeze_and_excitation.py:77: <font color="#CC0000"><b>error:</b></font> Cannot determine type of &apos;b&apos;
monai/networks/blocks/squeeze_and_excitation.py:77: <font color="#CC0000"><b>error:</b></font> Cannot determine type of &apos;c&apos;
monai/visualize/img2tensorboard.py:40: <font color="#CC0000"><b>error:</b></font> Tuple index out of range
<font color="#CC0000"><b>Found 29 errors in 11 files (checked 306 source files)</b></font>
</pre>
### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Thanks for your PR, don't worry about the typehints errors, I also encountered them locally and I am working on it to fix.

File: docs/source/networks.rst [hunk 1]
```diff
@@ -35,6 +35,21 @@ Blocks
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. autoclass:: SEBottleneck
+    :members:
+
+`Squeeze-and-Excitation Resnet Bottleneck`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. autoclass:: SEResNetBottleneck
+    :members:
+
+`Squeeze-and-Excitation Resnext Bottleneck`
```
Comment (Nic-Ma) [inline docs/source/networks.rst:48]: I feel it should be `Squeeze-and-Excitation ResneXt Bottleneck`

File: monai/networks/blocks/convolutions.py [hunk 2]
```diff
@@ -60,10 +61,11 @@ def __init__(
         out_channels: int,
         strides: int = 1,
         kernel_size: Union[Sequence[int], int] = 3,
-        act=Act.PRELU,
-        norm=Norm.INSTANCE,
-        dropout=None,
+        act: Optional[Union[tuple, str]] = Act.PRELU,
```
Comment (Nic-Ma) [inline monai/networks/blocks/convolutions.py:64]: I think it's not `Optional`, please double confirm.
Thanks.

File: monai/networks/blocks/convolutions.py [hunk 3]
```diff
@@ -60,10 +61,11 @@ def __init__(
...
+        norm: Optional[Union[tuple, str]] = Norm.INSTANCE,
```
Comment (Nic-Ma) [inline monai/networks/blocks/convolutions.py:65]: I think it's not `Optional`, please double confirm.
Thanks.

File: tests/test_senet.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    {"spatial_dims": 3, "in_ch": 2, "num_classes": 10},
+    torch.randn(5, 2, 64, 64, 64),
+    (5, 10),
+]
+
+
+class TestSENET(unittest.TestCase):
+    @parameterized.expand([TEST_CASE_1])
+    def test_senet154_shape(self, input_param, input_data, expected_shape):
+        net = senet154(**input_param)
```
Comment (Nic-Ma) [inline tests/test_senet.py:37]: There is a lot of duplicated code in this test file, I suggest to put `net` in `TEST_CASE`, something like:
```py
TEST_CASE_1 = [senet154(spatial_dims=3, in_ch=2, num_classes=10)]

TEST_CASE_2 = [se_resnet50(spatial_dims=3, in_ch=2, num_classes=10)]
... ...


class TestSENET(unittest.TestCase):
    @parameterized.expand([TEST_CASE_1, TEST_CASE_2, ... ...])
    def test_shape(self, net):
        input_data = torch.randn(5, 2, 64, 64, 64)
        expected_shape = (5, 10)
        net.eval()
        with torch.no_grad():
            result = net.forward(input_data)
            self.assertEqual(result.shape, expected_shape)
```
What do you think?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli , do you know why our CI server doesn't report the typehint errors as Yiheng said?
I can reproduce these errors locally and when I updated PyTorch docker to 20.06, it became another bunch of typehint errors.

Thanks.

File: monai/networks/blocks/squeeze_and_excitation.py [hunk 5]
```diff
@@ -183,10 +189,12 @@ def __init__(
...
         )
 
-        if in_channels != n_chns_3:  # in the case of residual chns and output chns doesn't match
+        self.project = project
+        if self.project is None and in_channels != n_chns_3:
             self.project = Conv[Conv.CONV, spatial_dims](in_channels, n_chns_3, kernel_size=1)
-        else:
-            self.project = None
+
+        relu_type: Type[nn.ReLU] = Act[Act.RELU]
```
Comment (wyli) [inline monai/networks/blocks/squeeze_and_excitation.py:196]: could you make this relu optional, and support different types such as PReLU?

**Comment by wyli** [comment]
> Hi @wyli , do you know why our CI server doesn't report the typehint errors as Yiheng said?
> I can reproduce these errors locally and when I updated PyTorch docker to 20.06, it became another bunch of typehint errors.
> 
> Thanks.

I think I came across a similar issue, but couldn't find out the root cause.
usually I run these commands and the errors disappear:
```bash
pip install -U -r requirements-dev.txt
./runtests.sh --clean
./runtests.sh --codeformat --nounittest
```

**Comment by Nic-Ma** [comment]
> > Hi @wyli , do you know why our CI server doesn't report the typehint errors as Yiheng said?
> > I can reproduce these errors locally and when I updated PyTorch docker to 20.06, it became another bunch of typehint errors.
> > Thanks.
> 
> I think I came across a similar issue, but couldn't find out the root cause.
> usually I run these commands and the errors disappear:
> 
> ```shell
> pip install -U -r requirements-dev.txt
> ./runtests.sh --clean
> ./runtests.sh --codeformat --nounittest
> ```

Hi @wyli ,

Your command `--nounittest` skipped the tests that generate these typehint errors?
I think these `typehint tests(mypy)` should run when PR submitted, it's part of our flake8 test, why it doesn't raise in CI?

Thanks.

File: docs/source/networks.rst (hunk 1 above)
Comment (yiheng-wang-nv) [reply docs/source/networks.rst:48]: Thanks, fixed.

File: monai/networks/blocks/convolutions.py (hunk 3 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/convolutions.py:65]: Yes, Norm layer should not be optional, I've changed. Whereas, activation layer should be optional according to 1) the original comments in the code, 2) sometimes we do not need this layer, or we may do some manipulations for features after the norm layer first, and then do the activations, just as what resnet's resudial block did  in the downsample part ([https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py](url))

File: tests/test_senet.py (hunk 4 above)
Comment (yiheng-wang-nv) [reply tests/test_senet.py:37]: Thanks, the code should be more clear, I've modified.

File: monai/networks/blocks/squeeze_and_excitation.py (hunk 5 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/squeeze_and_excitation.py:196]: Sure, I've modified.

**Pull request merged**